### PR TITLE
bn256: added consensys/gurvy bn256 implementation

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
This pr adds consensys' gurvy bn256 variant into the code for differential fuzzing.
Since it is specially built with the correct parameters we don't need to vendor it.

### CODE REVIEW

**Comment by holiman** [comment]
> For this we need to wait for ConsenSys/gurvy#17.

It's merged

**Comment by MariusVanDerWijden** [comment]
Well then I know what to work on next :D

File: crypto/bn256/bn256_fuzz.go [hunk 1]
```diff
@@ -2,14 +2,15 @@
 // Use of this source code is governed by a BSD-style license that can be found
 // in the LICENSE file.
 
-// +build gofuzz
+//
```
Comment (MariusVanDerWijden) [inline crypto/bn256/bn256_fuzz.go:5]: TODO: add the build flags back in

**Comment by holiman** [comment]
> The fuzzers only do essentially random fuzzing without structure which means ~ 1 in 80 million tests produces a valid input to add or mulexp, the pairing is probably even worse

Also, it's super-picky about input -- essentially disqualifyiing everything that is not exactly the correct size. A more forgiving implementation could instead read input in chunks of `required_size`, and loop over the chunks. Don't know many cases are discarded for the reason over being 'too large', but might be a couple

**Comment by holiman** [comment]
Maybe let's do something like this too: 
```diff
diff --git a/crypto/bn256/bn256_fuzz.go b/crypto/bn256/bn256_fuzz.go
index 6aa1421170..dda80ff149 100644
--- a/crypto/bn256/bn256_fuzz.go
+++ b/crypto/bn256/bn256_fuzz.go
@@ -8,42 +8,44 @@ package bn256
 
 import (
 	"bytes"
+	"fmt"
+	"io"
 	"math/big"
 
 	cloudflare "github.com/ethereum/go-ethereum/crypto/bn256/cloudflare"
 	google "github.com/ethereum/go-ethereum/crypto/bn256/google"
 )
 
-// FuzzAdd fuzzez bn256 addition between the Google and Cloudflare libraries.
-func FuzzAdd(data []byte) int {
-	// Ensure we have enough data in the first place
-	if len(data) != 128 {
-		return 0
+func getG1Points(input io.Reader) (*cloudflare.G1, *google.G1) {
+	xc, err := cloudflare.RandomG1()
+	if err != nil {
+		// insufficient input
+		return nil, nil
+	}
+	dat, err := xc.MarshalText()
+	if err != nil {
+		panic("marshalling failed")
 	}
-	// Ensure both libs can parse the first curve point
-	xc := new(cloudflare.G1)
-	_, errc := xc.Unmarshal(data[:64])
-
 	xg := new(google.G1)
-	_, errg := xg.Unmarshal(data[:64])
-
-	if (errc == nil) != (errg == nil) {
-		panic("parse mismatch")
-	} else if errc != nil {
-		return 0
+	if _, err := xg.Unmarshal(dat); err != nil {
+		panic(fmt.Sprintf("Could not marshal couldflare -> google:", err))
 	}
-	// Ensure both libs can parse the second curve point
-	yc := new(cloudflare.G1)
-	_, errc = yc.Unmarshal(data[64:])
+	return xc, xg
 
-	yg := new(google.G1)
-	_, errg = yg.Unmarshal(data[64:])
+}
 
-	if (errc == nil) != (errg == nil) {
-		panic("parse mismatch")
-	} else if errc != nil {
+// FuzzAdd fuzzez bn256 addition between the Google and Cloudflare libraries.
+func FuzzAdd(data []byte) int {
+	input := bytes.NewReader(data)
+	xc, xg := getG1Points(input)
+	if xc == nil {
 		return 0
 	}
+	yc, yg := getG1Points(input)
+	if yc == nil {
+		return 0
+	}
+	// Ensure both libs can parse the second curve point
 	// Add the two points and ensure they result in the same output
 	rc := new(cloudflare.G1)
 	rc.Add(xc, yc)
```

**Comment by holiman** [comment]
Or let me make an alternative PR about that first... ?

**Comment by MariusVanDerWijden** [comment]
I think that should be an alternative pr @holiman (that could get merged in first)
Gurvy is still pretty rough around the edges so I think we should give it another month or two before putting it in

**Comment by holiman** [comment]
https://github.com/ethereum/go-ethereum/pull/21815

**Comment by gbotrel** [comment]
@MariusVanDerWijden hi :) . `gurvy` on [develop](https://github.com/ConsenSys/gurvy/tree/develop) should be 100% compatible with google and cloudflare implementations, including subgroup checks during deserialization. 

We're making a pass on the `pairing` APIs so that they are easily compatible with the EVM precompiles .  Then we will merge in `master` and ping you (in few days).

**Comment by MariusVanDerWijden** [comment]
Perfect! thank you very much @gbotrel

**Review by DGKSK8LIFE — APPROVED** [review]

**Comment by MariusVanDerWijden** [comment]
One problem we currently face is that `gurvy` uses `go-ethereum` internally creating a dependency cycle which is technically not problematic. However it results in gurvy pulling in 116 additional dependencies. gurvy uses the `go-ethereum` packages `crypto/bn256/cloudflare` and `crypto/bn256/google` to test against in `bn256/interop_test.go`

**Comment by DGKSK8LIFE** [comment]
Hmm. Sounds like this is inherent to Gurvy, right? So seems like there's nothing you can really do here? @MariusVanDerWijden

**Comment by gbotrel** [comment]
@MariusVanDerWijden [this issue](https://github.com/ConsenSys/gurvy/issues/22) should be fixed, using gurvy@v0.3.7 should work, let me know if you have other issues, happy to help.

**Comment by holiman** [comment]
@MariusVanDerWijden is this still WIP or good to go?

**Comment by MariusVanDerWijden** [comment]
I think its good to go, will rebase once more

**Comment by gbotrel** [comment]
@MariusVanDerWijden can you make `gurvy` point to `v0.3.8` in the `go.mod`? @yelhousni merged some algorithmic / performance improvement to the pairing (& multi-pairing) (+ [benchmarks](https://hackmd.io/@zkteam/eccbench#BN254) )

Cheers,

File: go.mod [hunk 2]
```diff
@@ -3,26 +3,21 @@ module github.com/ethereum/go-ethereum
...
 	github.com/cloudflare/cloudflare-go v0.10.2-0.20190916151808-a80f83b9add9
+	github.com/consensys/gurvy v0.3.8
 	github.com/davecgh/go-spew v1.1.1
 	github.com/deckarep/golang-set v0.0.0-20180603214616-504e848d77ea
-	github.com/dlclark/regexp2 v1.2.0 // indirect
 	github.com/docker/docker v1.4.2-0.20180625184442-8e610b2b55bf
 	github.com/dop251/goja v0.0.0-20200721192441-a695b0cdd498
 	github.com/edsrzf/mmap-go v1.0.0
-	github.com/fatih/color v1.3.0
+	github.com/fatih/color v1.7.0
```
Comment (holiman) [inline go.mod:18]: These are probably not actual changes related to this pr, are they? Might need a rebase ?

**Review by holiman — APPROVED** [review]
LGTM but please look into the go.mod changes

File: go.mod (hunk 2 above)
Comment (MariusVanDerWijden) [reply go.mod:18]: It's super weird, whenever I run `go test ./...` it just updates the `go-mod` with some unrelated changes. I removed them now by hand

**Review by holiman — APPROVED** [review]
LGTM

**Pull request merged**